scanpy/external/tl/_harmony_timeseries.py

```diff
@@ -0,0 +1,213 @@
+
+    Harmony time series is a framework for data visualization, trajectory
+    detection and interpretation for scRNA-seq data measured at discrete
+    time points. Harmony constructs an augmented affinity matrix by augmenting
+    the kNN graph affinity matrix with mutually nearest neighbors between
+    successive time points. This augmented affinity matrix forms the basis for
+    generated a force directed layout for visualization and also serves as input
+    for computing the diffusion operator which can be used for trajectory
+    detection using **Palantir**.
+
+    More about **Palantir** can be found here:
+    https://github.com/dpeerlab/Palantir.
+
+    .. note::
+        More information and bug reports `here
+        <https://github.com/dpeerlab/Harmony>`__.
+
+    Parameters
+    ----------
+    data
+        Annotated data matrix of shape n_obs `×` n_vars. Rows correspond to
+        cells and columns to genes. Rows represent two or more time points,
+        where replicates of the same time point are consecutive in order.
+
+    tp
+        key name of observation annotation `.obs` representing time points
+
+    Returns
+    -------
+    Updates `.uns` with `timepoint_connections`
+
+    Example
+    -------
+
+    >>> import scanpy as sc
+    >>> import scanpy.external as sce
+
+    **Load** `AnnData`
+
+    A sample with real data is available `here
+    <https://github.com/dpeerlab/Harmony/tree/master/data>`_.
+
+    Random data sets of three time points with two replicates each:
+
+    >>> adata_ref = sc.datasets.pbmc3k()
+    >>> start = [596, 615, 1682, 1663, 1409, 1432]
+    >>> adatas = [adata_ref[i : i + 1000] for i in start]
+    >>> sample_names = [
+            "sa1_Rep1",
+            "sa1_Rep2",
+            "sa2_Rep1",
+            "sa2_Rep2",
+            "sa3_Rep1",
+            "sa3_Rep2",
+        ]
+    >>> timepoints = [i.split("_")[0] for i in sample_names]
+    >>> for ad, sn, tp in zip(adatas, sample_names, timepoints):
+            ad.obs["time_points"] = tp
+            ad.obs["sample_name"] = sn
+    >>> adata = adatas[0].concatenate(*adatas[1:], join="outer")
+
+    Normalize and filter for highly expressed genes
+
+    >>> sc.pp.normalize_total(adata, target_sum=10000)
+    >>> sc.pp.log1p(adata)
+    >>> sc.pp.highly_variable_genes(adata, n_top_genes=1000, subset=True)
+
+    Run harmony_timeseries
+
+    >>> d = sce.tl.harmony_timeseries(data=adata, tp="time_points")
+
+    **Harmony augmented affinity matrix**
+
+    >>> aug_aff, aff = d.harmony_timeseries.core.augmented_affinity_matrix(
+            data_df=adata.to_df(),
+            timepoints=adata.obs["time_points"],
+            timepoint_connections=adata.uns["timepoint_connections"],
+        )
+
+    **Visualization using force directed layouts**
+
+    >>> layout = d.harmony_timeseries.plot.force_directed_layout(
+            affinity_matrix=aug_aff, cell_names=adata.obs.index
+        )
```

Comment:
Is there any reason you need to return a class? Why not

1. Do both of those steps within `sce.tl.harmony_timeseries`
2. store the resulting matrices within `adata`

Reply:
Done!